DataFormats/Portable/interface/PortableCollectionCommon.h

```diff
@@ -0,0 +1,113 @@
+
+#ifndef DataFormats_Portable_interface_PortableCollectionCommon_h
+#define DataFormats_Portable_interface_PortableCollectionCommon_h
+
+template <auto Start, auto End, auto Inc, class F>
+constexpr void constexpr_for(F&& f) {
+  if constexpr (Start < End) {
+    f(std::integral_constant<decltype(Start), Start>());
+    constexpr_for<Start + Inc, End, Inc>(f);
+  }
+}
+
+template <std::size_t idx, typename T>
+struct CollectionLeaf {
+  CollectionLeaf() = default;
+  CollectionLeaf(std::byte* buffer, int32_t elements) : layout_(buffer, elements), view_(layout_) {}
+  template <std::size_t N>
+  CollectionLeaf(std::byte* buffer, std::array<int32_t, N> sizes) : layout_(buffer, sizes[idx]), view_(layout_) {
```

Comment:
would it be better to pass the array as a const reference ?
```suggestion
  CollectionLeaf(std::byte* buffer, std::array<int32_t, N> const& sizes) : layout_(buffer, sizes[idx]), view_(layout_) {
```